### ISSUE TITLE
Always remove deleted projects from working sets

### ISSUE THREAD

**Issue by HannesWell** [eclipse-platform/eclipse.platform.ui#2250]
- [x] I verified [latest Integration Build of Eclipse SDK](https://download.eclipse.org/eclipse/downloads/) doesn't provide the feature.

## Suggestion

If I delete a project from the disk via the package explorer it often happens it remains as closed project in the working set it was assigned too. This is especially annoying because I then cannot delete the project from the working set because the dialog then says that the project does not exist anymore and therefore cannot be deleted.
A workaround seems to be to explicitly remove the project from the working sets (e.g. in its properties).

Ideally the latter would always happen automatically when a project is deleted.

## Community

- [x] I understand suggesting an enhancement doesn't mandate anyone to implement it. Other contributors may consider this suggestion, or not, at their own convenience. The most efficient way to get it fixed is that I implement it myself and contribute it back as a good quality patch to the project.

**Comment by jukzi** [comment]
duplicate of https://github.com/eclipse-platform/eclipse.platform/issues/1074 ?

**Comment by iloveeclipse** [comment]
Interestingly, I always (since years) work with working sets and never needed to delete any project from working set after project deletion. 

I guess it's not about working sets, but about projects remaining "known" by the workspace after they were deleted outside of Eclipse and so shown in all "explorers"?

**Comment by HannesWell** [comment]
> duplicate of [eclipse-platform/eclipse.platform#1074](https://github.com/eclipse-platform/eclipse.platform/issues/1074) ?

Not exactly, but it happens in the same scenario. While it is probably makes sense to work on both at the same time, this is only about projects removed from the file-system staying in working sets.

> I guess it's not about working sets, but about projects remaining "known" by the workspace after they were deleted outside of Eclipse and so shown in all "explorers"?

As far as I can tell from my experiments the projects are not displayed when e.g. the `Package-Explorer` is configured to have project's as top level elements. But they remain present if `working sets` are selected as top level elements and I still can remove them from the configuration of the working set in its settings.

**Comment by jim-zy** [comment]
I'm working on this

**Comment by HannesWell** [comment]
> I'm working on this

Great. You are one of the students from CodeDays, aren't you?

Much success on this task and don't hesitate to ask for help if you need any.

**Comment by A-Lauper** [comment]
Hi, I am part of Jim’s student team with CodeDay.  Thank you for your words of encouragement.
I am having difficulty reproducing the issue as I understand it.

What I have tried:

Attempt 1
Set the Package Explorer top level elements to “working sets”.
Delete from disk a project assigned to a working set through the Package Explorer in Eclipse.
The project appears to have been removed from the working set.
Video: https://www.youtube.com/watch?v=uq35eJurQW8

Attempt 2
Delete a project from the system file explorer outside of Eclipse.
The project remains in the working set.
If I refresh the Package Explorer, it detects that the project has been deleted and prompts to delete the project from the workspace.
Video: https://www.youtube.com/watch?v=FeAUICRkyxY

I have attempted with both Java and Resource working sets and have had the same results.

To clarify, is the issue that projects are not automatically removed from the workspace when deleted outside of Eclipse, or is the issue directly related to when projects are deleted through the Package Explorer in Eclipse?

Build Id Information:
Eclipse Platform
Version: 2024-12 (4.34)
Build id: I20241016-2000
OS: Windows 11, v.10.0, x86_64 / win32
Java vendor: Oracle Corporation
Java runtime version: 17.0.11+7-LTS-207
Java version: 17.0.11

Thank you!

**Comment by jim-zy** [comment]
> > I'm working on this
> 
> Great. You are one of the students from CodeDays, aren't you?
> 
> Much success on this task and don't hesitate to ask for help if you need any.

Hi, we found the same issue when deleting plug-in projects from a plug-in working set, the project remains in the working set after deleted in the package explorer, did you encounter the issue with plug-in project? @HannesWell

**Comment by HannesWell** [comment]
Sorry for the delayed reply.

> > > I'm working on this
> > 
> > 
> > Great. You are one of the students from CodeDays, aren't you?
> > Much success on this task and don't hesitate to ask for help if you need any.
> 
> Hi, we found the same issue when deleting plug-in projects from a plug-in working set, the project remains in the working set after deleted in the package explorer, did you encounter the issue with plug-in project? @HannesWell

Great! Yes I encountered this problem with a Plug-in project (i.e. with a project that has PDE's nature `org.eclipse.pde.PluginNature` specified in the `.project` file.
I haven't tested it with projects that don't have that nature. Can you check if the nature makes a difference? If yes that could already be a strong hint for the underlying cause.

**Comment by jim-zy** [comment]
> Sorry for the delayed reply.
> 
> > > > I'm working on this
> > > 
> > > 
> > > Great. You are one of the students from CodeDays, aren't you?
> > > Much success on this task and don't hesitate to ask for help if you need any.
> > 
> > 
> > Hi, we found the same issue when deleting plug-in projects from a plug-in working set, the project remains in the working set after deleted in the package explorer, did you encounter the issue with plug-in project? @HannesWell
> 
> Great! Yes I encountered this problem with a Plug-in project (i.e. with a project that has PDE's nature `org.eclipse.pde.PluginNature` specified in the `.project` file. I haven't tested it with projects that don't have that nature. Can you check if the nature makes a difference? If yes that could already be a strong hint for the underlying cause.

Yes, the nature matters and a Java project could be deleted automatically from the working set. Also, I found that the PDE related problem you mentioned could be solved if you press F5 as the refresh button in the project explorer, then the project that was deleted from the package explorer will disappear. So this could be a UI problem.

**Comment by HannesWell** [comment]
I also did some experiments and for me the nature of the project didn't made a difference.

In general the modification of working-sets seems not to work well, even additions didn't always work in my workspace with many working sets and many projects and not the creation of the working sets themself was displayed in the Package Explorer immediately.

But since I couldn't create the working-set or project a second time, then the error was that it already exists, I agree that it's probably a UI problem.

One cause could maybe also be that the [org.eclipse.ui.navigator.resources.ProjectExplorer](https://github.com/eclipse-platform/eclipse.platform.ui/blob/9474c952a23a2c898ba25fd8a09d498434bca0af/bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/navigator/resources/ProjectExplorer.java#L85) provided by eclipse.platform.ui
and the [org.eclipse.jdt.internal.ui.packageview.PackageExplorerPart](https://github.com/eclipse-jdt/eclipse.jdt.ui/blob/324bdfd6be1c10820b9659e9190a158b9dda6cae/org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/packageview/PackageExplorerPart.java#L152) provided by eclipse.jdt.ui interfere somehow.

Maybe you can check those two classes if they might interfere or if they are maybe not correctly updated if the underlying resource model changes?

**Comment by freyajiang** [comment]
just discovered that deleting projects from the Package Explorer triggers the removeWorkingSet(IWorkingSet workingSet) function in the codebase. This might be a useful starting point for addressing the issue at hand

**Comment by nettozahler** [comment]
I would like to bring this forward because for me it's clearly an annoying bug that I can reproduce. @HannesWell or @jjohnstn would you like to step in to support me? I will start with a description of my setup and how to reproduce as well as some other observations.
I use the Eclipse "Java Browsing" perspective and have both views "Projects" (in the background) and "Project Explorer" open (not sure if this is relevant). I work with "Project Explorer" view and here the bug shows up. View settings are: Project Presentation = "Hierarchical" and Top Level Elements = "Working Sets". This is important. Have at least one self-defined Working Set present containing a project. Assert there is no "Other projects" working set yet.

**Bug 1:** Create a new Java project. **It does not appear** in the "Project Explorer" afterwards. But it exists because you cannot create a second Java project with same name. **Expectation**: the new project should appear immediately under an automatically created working set "Other projects".

To "**heal**" the wrong UI status you can restart the IDE **or** switch the setting "Top Level Elements" to "Projects" and then back to "Working Sets".

**Bug 2:** Delete the new  Java project in working set "Other projects" using the option "Delete project contents on disk".  **It does not disappear** in the "Project Explorer" afterwards. Instead it stays visible as _closed_ project.  **Expectation**: the deleted project should disappear immediately.

Again you can "**heal**" the wrong UI status as explained above.

Probably both bugs are related and can be fixed together. The "happy path" for the deletion problem is when you delete a Java project which is in a working set **other than** "Other projects". This works fine.

Some debugging has shown that for the **deletion problem** in `org.eclipse.jdt.internal.ui.workingsets.JavaWorkingSetUpdater.elementChanged(ElementChangedEvent)` the workingSet delta is not processed because the working set "Other projects" is not real but a dummy. Therefore no update mechanisms get triggered.

**Comment by nettozahler** [comment]
@jjohnstn would be great if you could support this. The bugs are annoying if projects are created/deleted frequently with the described settings. Thank you.

**Comment by jjohnstn** [comment]
@nettozahler I'll take a look.

**Comment by jjohnstn** [comment]
@nettozahler I posted a PR that updates the JavaWorkingSetUpdater to ensure that the Other Projects working set is processed in addition to the local list of working sets.  This seems to fix the problems you mention above.  Adding a new project adds it to Other Projects and it appears in all views.  Deleting from Other Projects working set removes the project from all the views.  Would you like to test it out and post to the PR?  If you need help setting up a workspace to test it, just post to the PR.

**Comment by nettozahler** [comment]
@jjohnstn  Thanks for the fix. I will test it out when I find some time in the next days.
